lib/FrontendTool/FrontendTool.cpp

```diff
@@ -769,43 +790,85 @@ static bool performCompile(CompilerInstance &Instance,
   assert(Action >= FrontendOptions::ActionType::EmitSILGen &&
          "All actions not requiring SILGen must have been handled!");
 
-  std::unique_ptr<SILModule> SM = Instance.takeSILModule();
-  // Records whether the SIL is directly computed from the AST we have, meaning
-  // that it will exactly match the source. It might not if, for instance, some
-  // of the inputs are SIB with extra explicit SIL.
-  auto astGuaranteedToCorrespondToSIL = false;
-  if (!SM) {
+  auto mod = Instance.getMainModule();
+  std::deque<PostSILGenInputs> PSGIs;
```

Comment:
Why *is* this a deque?

Reply:
To preserve order. The elements get moved-out for subsequent processing. Other options:

  - Use a vector and reverse before use
  - Use a vector and don't worry about preserving order
  - Use a vector and don't worry about cost of shifting elements down when moving first element out.
  - Use a linked list or some other container because deque is undesirable
  - Restructure subsequent callees to not take ownership, don't move, just pass references in order

I just use a deque when I want "vector, but with O(1) front removal"

Reply:
Ah, I didn't realize you were actually moving things out of it. I thought it was just iterated over later.

…is there any reason to move things out of it?

Reply:
The downstream subsystems take ownership of the SILModule(s) via unique_ptr<> and destroy them incrementally (in parallel, in fact, on separate threads) post-IRGen, while performLLVM is running. My guess is this helps a bit with limiting peak memory use in an MT scenario as well as helping keep the cores saturated with work, similar to how many GCs run sweep on a separate thread.

(and in general I think it's a good pattern to pass ownership rather than references, if we want to move towards further pipelining & parallelism, so I didn't really think it made sense to undo the current usage)